src/dbn_struct.jl

```diff
@@ -0,0 +1,174 @@
+"""
+    DBNVar(x::Symbol)
+    DBNVar{x::Symbol}()
+
+Reference to a named node in the POMDP or MDP dynamic Bayesian network (DBN).
+
+Note that `gen(::DBNVar, m, depargs..., rng)` always takes an argument for each dependency, `gen(::DBNOut, m, s, a, rng)` only takes `s` and `a` arguments (the inputs to the entire DBN).
+
+`DBNVar` is a "value type". See the documentation of `Val` for more conceptual details about value types.
+"""
+struct DBNVar{name} end
+
+DBNVar(name::Symbol) = DBNVar{name}()
+
+"""
+    DBNOut(x::Symbol)
+    DBNOut{x::Symbol}()
+    DBNOut(::Symbol, ::Symbol,...)
+    DBNOut{x::NTuple{N, Symbol}}()
+
+Reference to one or more named nodes in the POMDP or MDP dynamic Bayesian network (DBN).
+
+Note that `gen(::DBNOut, m, s, a, rng)` always takes `s` and `a` arguments (the inputs to the entire DBN) while `gen(::DBNVar, m, depargs..., rng)` takes a variable number of arguments (one for each dependency).
+
+`DBNOut` is a "value type". See the documentation of `Val` for more conceptual details about value types.
+"""
+struct DBNOut{names} end
+
+DBNOut(name::Symbol) = DBNOut{name}()
+DBNOut(names...) = DBNOut{names}()
+
+struct DBNDef{N<:NamedTuple}
+    nodes::N
+    deps::NamedTuple # this could be a parameter in the future
+end
+
+# todo DBNDef structure
+
+node(d::DBNDef, name::Symbol) = d.nodes[name]
+deps(d::DBNDef, name::Symbol) = d.deps[name]
+nodenames(d::DBNDef) = keys(d.nodes)
+@generated function add_node(d::DBNDef, n::DBNVar{name}, node, deps) where name
+    quote 
+        @assert !haskey(d.nodes, name) "DBNDef already has a node named :$name"
+        DBNDef(merge(d.nodes, ($name=node,)), merge(d.deps, ($name=deps,)))
+    end
+end
+
+function mdp_dbn()
+    DBNDef((s = InputDBNNode(),
+            a = InputDBNNode(),
+            sp = DistributionDBNNode(transition),
+            r = FunctionDBNNode(reward),
+           ),
+           (s = (),
+            a = (),
+            sp = (:s, :a),
+            r = (:s, :a, :sp),
+           )
+          )
+end
+
+function pomdp_dbn()
+    DBNDef((s = InputDBNNode(),
+            a = InputDBNNode(),
+            sp = DistributionDBNNode(transition),
+            o = DistributionDBNNode(observation),
+            r = FunctionDBNNode(reward),
+           ),
+           (s = (),
+            a = (),
+            sp = (:s, :a),
+            o = (:s, :a, :sp),
+            r = (:s, :a, :sp, :o),
+           )
+          )
+end
+
+"""
+    DBNStructure(::Type{M}) where M <: Union{MDP, POMDP}
+
+Trait of an MDP/POMDP type for describing the structure of the dynamic Baysian network.
+
+# Example
```

Comment:
Maybe you could use `jldoctest` here.

Reply:
yes! I just don't know how to do it

Reply:
Seems like you can just wrap it in a `jldoctest` code environment. [See here](https://juliadocs.github.io/Documenter.jl/stable/man/doctests/index.html). Potentially, you have be `using POMDPs` via [DocMeta](https://juliadocs.github.io/Documenter.jl/stable/man/doctests/index.html#Module-level-metadata-1). But maybe this is also not something to worry about now but rather in a separate PR for the entire repo.